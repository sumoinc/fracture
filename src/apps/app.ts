import { join } from "path";

import {
  AwsCdkTypeScriptApp,
  AwsCdkTypeScriptAppOptions,
} from "projen/lib/awscdk";
import { TypeScriptProject } from "projen/lib/typescript";
import { SetRequired } from "type-fest";
import {
  FractureSubProjectOptions,
  fractureProjectInit,
  fractureProjectOptions,
} from "../fracture-project";
import { Settings } from "../settings";
import { TurboRepo } from "../turborepo";
import { DeployJobOptions, Workflow } from "../workflows";

export type AppOptions = Omit<
  SetRequired<Partial<AwsCdkTypeScriptAppOptions>, "name">,
  "parent"
> &
  FractureSubProjectOptions;

export class App extends AwsCdkTypeScriptApp {
  /**
   *
   */
  readonly parent: TypeScriptProject;

  constructor(options: AppOptions) {
    // All apps must have a parent project.
    if (!options.parent) {
      throw new Error("all apps must have a parent project");
    }

    // grab settings from parent project
    const { defaultCdkVersion, appRoot } = Settings.of(options.parent);

    // make sure services is configured as a workspace
    TurboRepo.of(options.parent).addWorkspaceRoot(appRoot);

    // strip org scopes from folder structure, if given
    const subdir =
      options.name.split("/").length === 2
        ? options.name.split("/")[1]
        : options.name;

    // set up the outdir
    const outdir = join(appRoot, subdir);

    super({
      cdkVersion: defaultCdkVersion,
      // merge with global settings
      ...fractureProjectOptions({
        ...options,
        outdir,
      }),
      artifactsDirectory: join(outdir, "cdk.out"),
    });

    this.parent = options.parent;

    // init some common things we need here
    fractureProjectInit(this);
  }

  public deploy(
    options: Pick<DeployJobOptions, "branchPrefix" | "environment">
  ) {
    const branchPrefix = options.branchPrefix ?? "main";
    // add to deployment workflow
    return Workflow.deploy(this.parent).addDeployJob({
      ...options,
      branchPrefix,
      appName: this.name,
      deploySteps: [
        {
          name: "deploy",
          run: `npx aws-cdk deploy --require-approval never --app ${this.artifactsDirectory} *-${branchPrefix}-${options.environment.name}`,
          // inlineScript: $(PACKAGE_MANAGER.PACKAGE_MANAGER_TYPE) cdk deploy --require-approval never --app 'cdk.out/${{ parameters.target_environment }}' ${{ parameters.cdk_stack_names }}
        },
      ],
      artifactsDirectory: this.artifactsDirectory,
    });
  }
}
