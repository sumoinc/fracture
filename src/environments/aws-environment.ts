import { NodeProject } from "projen/lib/javascript";
import { ValueOf } from "type-fest";
import {
  AuthProviderType,
  Environment,
  EnvironmentOptions,
} from "./environment";
import { Settings } from "../fracture-settings";

export const AwsRegion = {
  US_EAST_1: "us-east-1",
  US_EAST_2: "us-east-2",
  US_WEST_1: "us-west-1",
  US_WEST_2: "us-west-2",
} as const;

export interface AwsEnvironmentOptions extends EnvironmentOptions {
  /**
   * Account number for this environment.
   */
  readonly accountNumber: string;

  /**
   * Region for this envirnoment.
   *
   * @default - uses default from Settings()
   */
  readonly region?: ValueOf<typeof AwsRegion>;

  /**
   * Type of auth provider to use in this environment
   *
   * @default - uses default from Settings()
   */
  readonly authProviderType?: ValueOf<typeof AuthProviderType>;

  /**
   * The role name that shuld ber used when constructing the OIDC role's ARN.
   *
   * @default - uses default from Settings()
   */
  readonly gitHubDeploymentOIDCRoleName?: string;

  /**
   * The duration of the role session in seconds.
   *
   * @default - uses default from Settings()
   */
  readonly gitHubDeploymentOIDCRoleDurationSeconds?: number;
}

export class AwsEnvironment extends Environment {
  /**
   * Account for this environment.
   */
  public readonly accountNumber: string;

  /**
   * Region for this envirnoment.
   *
   * @default - uses default from Settings()
   */
  public readonly region: ValueOf<typeof AwsRegion>;

  /**
   * Type of auth provider to use in this environment
   *
   * @default - uses default from Settings()
   */
  readonly authProviderType: ValueOf<typeof AuthProviderType>;

  /**
   * The role name that shuld ber used when constructing the OIDC role's ARN.
   *
   * @default - uses default from Settings()
   */
  readonly gitHubDeploymentOIDCRoleName: string;

  /**
   * The duration of the role session in seconds.
   *
   * @default - uses default from Settings()
   */
  readonly gitHubDeploymentOIDCRoleDurationSeconds: number;

  constructor(
    public readonly project: NodeProject,
    options: AwsEnvironmentOptions
  ) {
    super(project, options);

    // grab some core defaults
    const {
      defaultRegion,
      gitHubDeploymentOIDCRoleName,
      gitHubDeploymentOIDCRoleDurationSeconds,
    } = Settings.of(project);

    // defaults
    this.accountNumber = options.accountNumber;
    this.region = options.region ?? defaultRegion;
    this.authProviderType =
      options.authProviderType ?? AuthProviderType.AWS_GITHUB_OIDC;
    this.gitHubDeploymentOIDCRoleName =
      options.gitHubDeploymentOIDCRoleName ?? gitHubDeploymentOIDCRoleName;
    this.gitHubDeploymentOIDCRoleDurationSeconds =
      options.gitHubDeploymentOIDCRoleDurationSeconds ??
      gitHubDeploymentOIDCRoleDurationSeconds;
  }

  // public get authProvider(): AuthProvider {
  //   return AuthProvider.fromAwsEnvironment(this.project, this);
  // }
}
