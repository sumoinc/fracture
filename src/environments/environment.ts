import { paramCase } from "change-case";
import { Component } from "projen";
import { NodeProject } from "projen/lib/javascript";
import { AuthProvider } from "../workflows/auth-provider";

export const EnvironmentType = {
  AWS: "AWS",
  NETLIFY: "NETLIFY",
  VERCEL: "VERCEL",
} as const;

export interface EnvironmentOptions {
  /**
   * Friendly name for environment
   */
  readonly name: string;
}

export class Environment extends Component {
  /**
   * Returns a environment by name, or undefined if it doesn't exist
   */
  public static byName(
    project: NodeProject,
    name: string
  ): Environment | undefined {
    const isDefined = (c: Component): c is Environment =>
      c instanceof Environment && c.name === name;
    return project.components.find(isDefined);
  }

  /**
   * Friendly name for environment
   */
  public readonly name: string;

  constructor(
    public readonly project: NodeProject,
    options: EnvironmentOptions
  ) {
    /***************************************************************************
     * Check Duplicates
     **************************************************************************/

    const name = paramCase(options.name);

    if (Environment.byName(project, name)) {
      throw new Error(`Duplicate environment name "${name}".`);
    }

    super(project);

    this.name = options.name;
  }

  public get authProvider(): AuthProvider {
    throw new Error("Not implemented, override in subclass.");
  }
}
