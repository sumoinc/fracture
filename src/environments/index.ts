export * from "./environment";
export * from "./region";
