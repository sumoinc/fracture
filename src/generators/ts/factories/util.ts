import { ValueOf } from "type-fest";
import { ResourceAttributeType } from "../../../services";

export const JsType = {
  STRING: "string",
  NUMBER: "number",
  BOOLEAN: "boolean",
  ARRAY: "array",
  MAP: "map",
  OTHER: "other",
} as const;

export const jsType = (
  resourceAttributeType: ValueOf<typeof ResourceAttributeType> | string
) => {
  switch (resourceAttributeType) {
    case ResourceAttributeType.GUID:
    case ResourceAttributeType.STRING:
    case ResourceAttributeType.EMAIL:
    case ResourceAttributeType.PHONE:
    case ResourceAttributeType.URL:
    case ResourceAttributeType.DATE:
    case ResourceAttributeType.TIME:
    case ResourceAttributeType.DATE_TIME:
    case ResourceAttributeType.JSON:
    case ResourceAttributeType.IPADDRESS:
      return JsType.STRING;
    case ResourceAttributeType.INT:
    case ResourceAttributeType.FLOAT:
    case ResourceAttributeType.TIMESTAMP:
    case ResourceAttributeType.COUNT:
    case ResourceAttributeType.AVERAGE:
    case ResourceAttributeType.SUM:
      return JsType.NUMBER;
    case ResourceAttributeType.BOOLEAN:
      return JsType.BOOLEAN;
    case ResourceAttributeType.ARRAY:
      return JsType.ARRAY;
    case ResourceAttributeType.MAP:
      return JsType.MAP;
    default:
      return JsType.OTHER;
  }
};
