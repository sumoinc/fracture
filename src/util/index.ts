export * from "./determineBuildId";
export * from "./test-util";
