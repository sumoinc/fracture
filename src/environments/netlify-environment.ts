import { NodeProject } from "projen/lib/javascript";
import { ValueOf } from "type-fest";
import { Environment, EnvironmentOptions } from "./environment";
import { Settings } from "../core/fracture-settings";
import { AuthProvider, AuthProviderType } from "../workflows/auth-provider";

export interface NetlifyEnvironmentOptions extends EnvironmentOptions {
  /**
   * Github Secret name for the NETLIFY_AUTH_TOKEN
   *
   * @default NETLIFY_AUTH_TOKEN
   */
  readonly authTokenSecretName?: string;

  /**
   * Github Secret name for the NETLIFY_SITE_ID
   *
   * @default NETLIFY_SITE_ID
   */
  readonly siteIdSecretName?: string;
}

export class NetlifyEnvironment extends Environment {
  /**
   * Github Secret name for the NETLIFY_AUTH_TOKEN
   *
   * @default NETLIFY_AUTH_TOKEN
   */
  readonly authTokenSecretName: string;

  /**
   * Github Secret name for the NETLIFY_SITE_ID
   *
   * @default NETLIFY_SITE_ID
   */
  readonly siteIdSecretName: string;

  constructor(
    public readonly project: NodeProject,
    options: NetlifyEnvironmentOptions
  ) {
    super(project, options);

    // defaults
    this.authTokenSecretName =
      options.authTokenSecretName ?? "NETLIFY_AUTH_TOKEN";
    this.siteIdSecretName = options.siteIdSecretName ?? "NETLIFY_SITE_ID";
  }

  /*
  public get authProvider(): AuthProvider {
    return AuthProvider.fromNetlifyEnvironment(this.project, this);
  }
  */
}
