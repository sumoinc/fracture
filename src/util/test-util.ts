import { Project } from "projen";
import { synthSnapshot } from "projen/lib/util/synth";

export const TEST_ACCOUNT_ONE = "000000000000";
export const TEST_ORG_ONE = "org-123456";
export const TEST_REGION_ONE = "us-east-1";

/**
 *
 * Helper to generate and return one file from fracture.
 * Useful when trying to test the final output of generated files.
 *
 * @param fracture
 * @param filepath
 * @returns collection of files starting with supplied filepath
 */
export const synthFiles = (project: Project, filepath: string = ""): any => {
  const snapshot = synthSnapshot(project);
  // console.log(Object.keys(snapshot));
  const filtered = Object.keys(snapshot)
    .filter((path) => path.startsWith(filepath))
    .reduce((obj, key) => {
      return {
        ...obj,
        [key]: snapshot[key],
      };
    }, {} as { [key: string]: any });
  return filtered;
};

export const synthFile = (project: Project, filepath: string): string => {
  const files = synthFiles(project, filepath);

  if (files.length > 1) {
    throw new Error(`more than one file matched filepath "${filepath}"`);
  }

  if (files.length === 0) {
    throw new Error(`No files matched filepath "${filepath}"`);
  }

  return files[filepath];
};

// /**
//  * Builds a simple typescript app project as a test harness
//  */
// export const testProject = () => {
//   return new Project({
//     name: "my-project",
//   });
// };

// export const testNodeProject = (options: Partial<NodeProjectOptions> = {}) => {
//   return new NodeProject({
//     name: "my-project",
//     defaultReleaseBranch: "main",
//     packageManager: NodePackageManager.PNPM,
//     pnpmVersion: "8",
//     ...options,
//   });
// };

// /**
//  * Builds a simple typescript app project as a test harness
//  */
// export const testFractureProject = (
//   options: Partial<FractureProjectOptions> = {}
// ) => {
//   const fp = new FractureProject({
//     name: "my-project",
//     ...options,
//   });

//   /*
//   fp.buildEnvironment = new AwsEnvironment(fp, {
//     name: "test",
//     account: TEST_ACCOUNT_ONE,
//     region: TEST_REGION_ONE,
//     primaryHostedZoneId: "foo",
//     primaryDomainName: "bar",
//   });
//   */

//   return fp;
// };

// /**
//  * Builds a simple data service as a test harness
//  */
// export const testApp = () => {
//   return new App({
//     parent: testFractureProject(),
//     name: "my-app",
//   });
// };

// /**
//  * Builds a simple data service as a test harness
//  */
// export const testDataService = () => {
//   return new DataService(testApp(), {
//     name: "my-service",
//   });
// };

// /**
//  * Build a simple resource to test with
//  */
// export const testResource = () => {
//   const service = testDataService();
//   return new Resource(service.app, {
//     service,
//     name: "my-resource",
//   });
// };

// /**
//  * A more complicated resource configuration
//  */
// export const complexService = () => {
//   const service = testDataService();

//   const person = new Resource(service.app, { service, name: "person" });

//   const club = new Resource(service.app, {
//     service,
//     name: "club",
//     comments: ["A club"],
//     attributeOptions: [{ name: "name" }],
//   });

//   person.addAttribute({
//     name: "first-name",
//     shortName: "fn",
//   });
//   person.addAttribute({
//     name: "last-name",
//     shortName: "ln",
//   });
//   person.addAttribute({
//     name: "height",
//     shortName: "ht",
//     type: ResourceAttributeType.INT,
//   });
//   person.addAttribute({
//     name: "is-frog",
//     shortName: "if",
//     type: ResourceAttributeType.BOOLEAN,
//     comments: ["Is this person a frog?"],
//   });
//   person.addAttribute({
//     name: "fav-colors",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: ResourceAttributeType.STRING,
//     comments: ["List opf favorite colors"],
//   });

//   // one specific club
//   person.addAttribute({
//     name: "one-club",
//     type: club,
//   });
//   // an array of clubs
//   person.addAttribute({
//     name: "array-of-clubs",
//     shortName: "ad",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: club,
//   });

//   // an array of persons
//   club.addAttribute({
//     name: "array-of-persons",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: person,
//   });

//   // operation
//   Operation.create(person);
//   Operation.read(person);
//   Operation.update(person);
//   Operation.delete(person);

//   return service;
// };

// /**
//  * Build a simple resource attribute to test with
//  */
// export const testResourceAttribute = () => {
//   const service = testDataService();
//   const resource = testResource();
//   return new ResourceAttribute(service.app, {
//     name: "my-attr",
//     resource,
//   });
// };

// /**
//  * Build a simple structure to test with
//  */
// export const testStructure = () => {
//   const service = testDataService();
//   return new Structure(service.app, { name: "my-structure" });
// };

// /**
//  * Build a simple operation to test with
//  */
// export const testOperation = () => {
//   const resource = testResource();
//   return new Operation(resource.service.app, {
//     name: "my-operation",
//     resource,
//   });
// };

// /**
//  * Build a nuxt site to test with
//  */
// export const testNuxtJsSite = () => {
//   return new NuxtJsSite({
//     parent: testFractureProject(),
//     name: "my-site",
//   });
// };
// export const testServiceTwoRelatedResources = () => {
//   const service = testDataService();
//   const person = new Resource(service.app, {
//     service,
//     name: "person",
//     attributeOptions: [{ name: "name", shortName: "n" }],
//   });
//   const club = new Resource(service.app, {
//     service,
//     name: "club",
//     attributeOptions: [{ name: "name", shortName: "n" }],
//   });
//   club.addAttribute({
//     name: "people",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: person,
//   });
//   person.addAttribute({
//     name: "clubs",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: club,
//   });
//   return { service, person, club };
// };
