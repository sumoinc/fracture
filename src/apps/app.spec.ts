import { App } from "./app";
import { AwsEnvironment } from "../environments";
import { synthFile, testApp } from "../util/test-util";

describe("success conditions", () => {
  test("Smoke test", () => {
    expect(testApp()).toBeTruthy();
  });
});

describe("validate generated project files", () => {
  test(".projen/tasks.json", () => {
    const app = testApp();
    // make sure deployment tasks get added correctly
    const usEast = new AwsEnvironment(app.parent, {
      name: "us-east",
      account: "0000000000",
    });
    // deploy both apps
    app.deploy({
      branchPrefix: "feature",
      environment: usEast,
    });

    const content = synthFile(app, ".projen/tasks.json");
    expect(content).toBeTruthy();
    expect(content).toMatchSnapshot();
    //console.log(JSON.stringify(content, null, 2));
  });

  test("turbo.json", () => {
    const content = synthFile(testApp().parent, "turbo.json");
    expect(content).toBeTruthy();
    expect(content).toMatchSnapshot();
    // console.log(JSON.stringify(content, null, 2));
  });

  test(".github/workflows/deploy.yml", () => {
    const appOne = testApp();
    const appTwo = new App({
      parent: appOne.parent,
      name: "@scope/my-other-app",
    });
    const usEast = new AwsEnvironment(appOne.parent, {
      name: "us-east",
      account: "0000000000",
    });

    // deploy both apps
    appOne.deploy({
      branchPrefix: "feature",
      environment: usEast,
    });
    appTwo.deploy({
      branchPrefix: "feature",
      environment: usEast,
    });

    const content = synthFile(appOne.parent, `.github/workflows/deploy.yml`);
    expect(content).toBeTruthy();
    expect(content).toMatchSnapshot();
    // console.log(content);
  });
});
