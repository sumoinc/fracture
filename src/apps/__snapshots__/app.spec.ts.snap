// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate generated project files .projen/tasks.json 1`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
  "env": Object {
    "PATH": "$(pnpm -c exec \\"node --print process.env.PATH\\")",
  },
  "tasks": Object {
    "build": Object {
      "description": "Full release build",
      "name": "build",
      "steps": Array [
        Object {
          "spawn": "pre-compile",
        },
        Object {
          "spawn": "compile",
        },
        Object {
          "spawn": "post-compile",
        },
        Object {
          "spawn": "test",
        },
        Object {
          "spawn": "package",
        },
      ],
    },
    "bundle": Object {
      "description": "Prepare assets",
      "name": "bundle",
    },
    "compile": Object {
      "description": "Only compile",
      "name": "compile",
    },
    "default": Object {
      "description": "Synthesize project files",
      "name": "default",
    },
    "deploy": Object {
      "description": "Deploys your CDK app to the AWS cloud",
      "name": "deploy",
      "steps": Array [
        Object {
          "exec": "cdk deploy",
          "receiveArgs": true,
        },
      ],
    },
    "destroy": Object {
      "description": "Destroys your cdk app in the AWS cloud",
      "name": "destroy",
      "steps": Array [
        Object {
          "exec": "cdk destroy",
          "receiveArgs": true,
        },
      ],
    },
    "diff": Object {
      "description": "Diffs the currently deployed app against your code",
      "name": "diff",
      "steps": Array [
        Object {
          "exec": "cdk diff",
        },
      ],
    },
    "eslint": Object {
      "description": "Runs eslint against the codebase",
      "name": "eslint",
      "steps": Array [
        Object {
          "exec": "eslint --ext .ts,.tsx --fix --no-error-on-unmatched-pattern src test build-tools",
        },
      ],
    },
    "install": Object {
      "description": "Install project dependencies and update lockfile (non-frozen)",
      "name": "install",
      "steps": Array [
        Object {
          "exec": "pnpm i --no-frozen-lockfile",
        },
      ],
    },
    "install:ci": Object {
      "description": "Install project dependencies using frozen lockfile",
      "name": "install:ci",
      "steps": Array [
        Object {
          "exec": "pnpm i --frozen-lockfile",
        },
      ],
    },
    "package": Object {
      "description": "Creates the distribution package",
      "name": "package",
    },
    "post-compile": Object {
      "description": "Runs after successful compilation",
      "name": "post-compile",
      "steps": Array [
        Object {
          "spawn": "synth:silent",
        },
      ],
    },
    "post-upgrade": Object {
      "description": "Runs after upgrading dependencies",
      "name": "post-upgrade",
    },
    "pre-compile": Object {
      "description": "Prepare the project for compilation",
      "name": "pre-compile",
    },
    "synth": Object {
      "description": "Synthesizes your cdk app into cdk.out",
      "name": "synth",
      "steps": Array [
        Object {
          "exec": "cdk synth",
        },
      ],
    },
    "synth:silent": Object {
      "description": "Synthesizes your cdk app into cdk.out and suppresses the template in stdout (part of \\"yarn build\\")",
      "name": "synth:silent",
      "steps": Array [
        Object {
          "exec": "cdk synth -q",
        },
      ],
    },
    "test": Object {
      "description": "Run tests",
      "name": "test",
      "steps": Array [
        Object {
          "exec": "jest --passWithNoTests --updateSnapshot",
          "receiveArgs": true,
        },
        Object {
          "spawn": "eslint",
        },
      ],
    },
    "test:watch": Object {
      "description": "Run jest in watch mode",
      "name": "test:watch",
      "steps": Array [
        Object {
          "exec": "jest --watch",
        },
      ],
    },
    "upgrade": Object {
      "description": "upgrade dependencies",
      "env": Object {
        "CI": "0",
      },
      "name": "upgrade",
      "steps": Array [
        Object {
          "exec": "pnpm update npm-check-updates",
        },
        Object {
          "exec": "npm-check-updates --upgrade --target=minor --peer --dep=dev,peer,prod,optional --filter=@types/jest,@types/node,@typescript-eslint/eslint-plugin,@typescript-eslint/parser,aws-cdk,esbuild,eslint-config-prettier,eslint-import-resolver-node,eslint-import-resolver-typescript,eslint-plugin-import,eslint-plugin-prettier,eslint,jest,jest-junit,npm-check-updates,prettier,projen,ts-jest,ts-node,typescript,aws-cdk-lib,constructs",
        },
        Object {
          "exec": "pnpm i --no-frozen-lockfile",
        },
        Object {
          "exec": "pnpm update @types/jest @types/node @typescript-eslint/eslint-plugin @typescript-eslint/parser aws-cdk esbuild eslint-config-prettier eslint-import-resolver-node eslint-import-resolver-typescript eslint-plugin-import eslint-plugin-prettier eslint jest jest-junit npm-check-updates prettier projen ts-jest ts-node typescript aws-cdk-lib constructs",
        },
        Object {
          "exec": "npx projen",
        },
        Object {
          "spawn": "post-upgrade",
        },
      ],
    },
    "watch": Object {
      "description": "Watches changes in your source code and rebuilds and deploys to the current account",
      "name": "watch",
      "steps": Array [
        Object {
          "exec": "cdk deploy --hotswap",
        },
        Object {
          "exec": "cdk watch",
        },
      ],
    },
  },
}
`;

exports[`validate generated project files turbo.json 1`] = `
Object {
  "$schema": "https://turborepo.org/schema.json",
  "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run \\"npx projen\\".",
  "pipeline": Object {
    "eslint": Object {
      "cache": false,
      "dependsOn": Array [
        "^eslint",
      ],
    },
    "turbo:build": Object {
      "dependsOn": Array [],
      "outputMode": "new-only",
    },
  },
}
`;
